In [4]:
# Original cell content was incomplete. No action needed as the necessary imports are in cell 27d65bf3.

## Step 1: Import Libraries

In [5]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

## Step 2: Load Dataset
The dataset is already loaded in the previous steps.

## Step 3: Explore the Dataset
We will check for missing values, data types, and column names.

In [6]:
print("\nMissing values per column:")
display(housing.frame.isnull().sum())


Missing values per column:


,0
MedInc,0
HouseAge,0
AveRooms,0
AveBedrms,0
Population,0
AveOccup,0
Latitude,0
Longitude,0
MedHouseVal,0


In [7]:
print("\nData types and non-null counts:")
housing.frame.info()


Data types and non-null counts:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


In [8]:
print("\nColumn names:")
display(housing.frame.columns)


Column names:


Index(['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup',
       'Latitude', 'Longitude', 'MedHouseVal'],
      dtype='object')

## Step 4: Select Features and Target
We will define our features (X) and target (y) for the model.

In [9]:
X = housing.frame.drop(columns=['MedHouseVal'])
y = housing.frame['MedHouseVal']

print("Features (X) shape:", X.shape)
print("Target (y) shape:", y.shape)

Features (X) shape: (20640, 8)
Target (y) shape: (20640,)


## Step 5: Split the Dataset
We will split the data into training and testing sets (80% training, 20% testing).

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (16512, 8)
X_test shape: (4128, 8)
y_train shape: (16512,)
y_test shape: (4128,)


## Step 6: Train Linear Regression Model
Now, we will initialize and train a Linear Regression model.

In [11]:
model = LinearRegression()
model.fit(X_train, y_train)

print("Model training complete.")

Model training complete.


## Step 7: Make Predictions
We will use the trained model to make predictions on the test set.

In [12]:
y_pred = model.predict(X_test)

print("Predictions made on the test set.")

Predictions made on the test set.


## Step 8: Evaluate the Model
We will evaluate the model's performance using Root Mean Squared Error (RMSE).

In [13]:
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"Root Mean Squared Error (RMSE): {rmse:.4f}")

print("\nExplanation:\nLower RMSE values indicate better model performance, as it means the predictions are closer to the actual values. RMSE gives us an idea of the average magnitude of the errors in our predictions, in the same units as the target variable (MedHouseVal, which is in hundreds of thousands of dollars).")

Root Mean Squared Error (RMSE): 0.7456

Explanation:
Lower RMSE values indicate better model performance, as it means the predictions are closer to the actual values. RMSE gives us an idea of the average magnitude of the errors in our predictions, in the same units as the target variable (MedHouseVal, which is in hundreds of thousands of dollars).


In [1]:
import pandas as pd
from sklearn.datasets import fetch_california_housing

# Load the California Housing dataset
housing = fetch_california_housing(as_frame=True)


### Dataset Description


In [2]:
print(housing.DESCR)


.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

:Number of Instances: 20640

:Number of Attributes: 8 numeric, predictive attributes and the target

:Attribute Information:
    - MedInc        median income in block group
    - HouseAge      median house age in block group
    - AveRooms      average number of rooms per household
    - AveBedrms     average number of bedrooms per household
    - Population    block group population
    - AveOccup      average number of household members
    - Latitude      block group latitude
    - Longitude     block group longitude

:Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,000).

This dataset was derived from the 1990 U.S. census, using one row per ce

### First 5 rows of the dataset


In [3]:
display(housing.frame.head())


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


## Analyzing the Titanic Dataset (Classification)

### Step 1: Load Dataset (Titanic)
We will load the Titanic dataset using `seaborn`.

In [14]:
# Load the Titanic dataset
titanic_df = sns.load_dataset('titanic')

print("Titanic dataset loaded successfully.")

Titanic dataset loaded successfully.


### Step 2: Data Cleaning (Titanic)
We need to handle missing values, convert categorical variables, and drop unnecessary columns to prepare the data for modeling.

In [15]:
# Display initial info to identify missing values and data types
print("\nInitial info for Titanic dataset:")
titanic_df.info()

print("\nMissing values before cleaning:")
display(titanic_df.isnull().sum())


Initial info for Titanic dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB

Missing values before cleaning:


,0
survived,0
pclass,0
sex,0
age,177
sibsp,0
parch,0
fare,0
embarked,2
class,0
who,0


In [16]:
# Handle missing values
# 'age': fill with median
titanic_df['age'].fillna(titanic_df['age'].median(), inplace=True)

# 'embarked': fill with the most frequent value (mode)
titanic_df['embarked'].fillna(titanic_df['embarked'].mode()[0], inplace=True)

# 'deck': this column has too many missing values (over 77%), so it's best to drop it.
titanic_df.drop(columns=['deck'], inplace=True)

# Drop other columns that are not useful for prediction or are redundant
# 'cabin' is similar to 'deck' with many missing values
# 'fare' is already numerical
# 'name', 'ticket', 'pclass' are not directly useful as features (pclass information is captured by 'class')
# 'alive', 'adult_male', 'who', 'alone' are often derived from other columns or less direct than 'sex', 'age', etc.
# 'embark_town' is redundant with 'embarked'
titanic_df.drop(columns=['alone', 'embark_town', 'class', 'adult_male', 'who', 'pclass', 'ticket', 'fare', 'parch', 'sibsp', 'alive', 'cabin', 'name'], inplace=True, errors='ignore')

# Convert categorical features to numerical using one-hot encoding
titanic_df = pd.get_dummies(titanic_df, columns=['sex', 'embarked'], drop_first=True)


print("\nTitanic dataset after cleaning and preprocessing (first 5 rows):")
display(titanic_df.head())

print("\nMissing values after cleaning:")
display(titanic_df.isnull().sum())


Titanic dataset after cleaning and preprocessing (first 5 rows):


/tmp/ipykernel_374/198954993.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  titanic_df['age'].fillna(titanic_df['age'].median(), inplace=True)
/tmp/ipykernel_374/198954993.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplac

,survived,age,sex_male,embarked_Q,embarked_S
0,0,22.0,True,False,True
1,1,38.0,False,False,False
2,1,26.0,False,False,True
3,1,35.0,False,False,True
4,0,35.0,True,False,True



Missing values after cleaning:


,0
survived,0
age,0
sex_male,0
embarked_Q,0
embarked_S,0


### Step 3: Select Features and Target (Titanic)
We will define our features (X) and target (y) for the classification model. The target will be 'survived'.

In [17]:
X_titanic = titanic_df.drop(columns=['survived'])
y_titanic = titanic_df['survived']

print("Features (X_titanic) shape:", X_titanic.shape)
print("Target (y_titanic) shape:", y_titanic.shape)

Features (X_titanic) shape: (891, 4)
Target (y_titanic) shape: (891,)


### Step 4: Split the Dataset (Titanic)
We will split the Titanic data into training and testing sets (80% training, 20% testing).

In [18]:
X_train_titanic, X_test_titanic, y_train_titanic, y_test_titanic = train_test_split(X_titanic, y_titanic, test_size=0.2, random_state=42)

print("X_train_titanic shape:", X_train_titanic.shape)
print("X_test_titanic shape:", X_test_titanic.shape)
print("y_train_titanic shape:", y_train_titanic.shape)
print("y_test_titanic shape:", y_test_titanic.shape)

X_train_titanic shape: (712, 4)
X_test_titanic shape: (179, 4)
y_train_titanic shape: (712,)
y_test_titanic shape: (179,)


### Step 5: Train Logistic Regression Model (Titanic)
For this classification problem, we will use `LogisticRegression`.

In [19]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Initialize and train the Logistic Regression model
# Increased max_iter for convergence with some datasets
model_titanic = LogisticRegression(max_iter=1000, random_state=42)
model_titanic.fit(X_train_titanic, y_train_titanic)

print("Logistic Regression model training complete for Titanic dataset.")

Logistic Regression model training complete for Titanic dataset.


### Step 6: Make Predictions (Titanic)
We will use the trained model to make predictions on the test set.

In [20]:
y_pred_titanic = model_titanic.predict(X_test_titanic)

print("Predictions made on the Titanic test set.")

Predictions made on the Titanic test set.


### Step 7: Evaluate the Model (Titanic)
We will evaluate the model's performance using Accuracy Score and a Classification Report.

In [21]:
accuracy_titanic = accuracy_score(y_test_titanic, y_pred_titanic)
print(f"Accuracy for Titanic Logistic Regression model: {accuracy_titanic:.4f}")

print("\nClassification Report for Titanic Logistic Regression model:")
print(classification_report(y_test_titanic, y_pred_titanic))

print("\nExplanation:\n")
print("Higher accuracy indicates a better classification model. It represents the proportion of correctly predicted instances out of the total instances.")
print("The classification report provides more detailed metrics like precision, recall, and F1-score for each class (survived/not survived), which are crucial for understanding the model's performance on imbalanced datasets or when different types of errors have different costs.")

Accuracy for Titanic Logistic Regression model: 0.7821

Classification Report for Titanic Logistic Regression model:
              precision    recall  f1-score   support

           0       0.80      0.84      0.82       105
           1       0.75      0.70      0.73        74

    accuracy                           0.78       179
   macro avg       0.78      0.77      0.77       179
weighted avg       0.78      0.78      0.78       179


Explanation:

Higher accuracy indicates a better classification model. It represents the proportion of correctly predicted instances out of the total instances.
The classification report provides more detailed metrics like precision, recall, and F1-score for each class (survived/not survived), which are crucial for understanding the model's performance on imbalanced datasets or when different types of errors have different costs.


## Evaluation Report

### 1. Problem Statement

**California Housing Dataset (Regression):** The goal was to predict the median house value in various districts of California based on several features like median income, house age, number of rooms, population, etc.

**Titanic Dataset (Classification):** The objective was to build a classification model to predict whether a passenger survived the Titanic shipwreck based on features such as age, sex, and embarkation point.

### 2. Datasets Used

**California Housing Dataset:**
*   **Source:** `sklearn.datasets.fetch_california_housing` (derived from the 1990 U.S. census).
*   **Characteristics:** 20640 instances, 8 numeric predictive attributes, and one target variable (Median House Value).
*   **Missing Values:** None.

**Titanic Dataset:**
*   **Source:** `seaborn.load_dataset('titanic')`.
*   **Characteristics:** 891 instances, a mix of numerical and categorical features, and a binary target variable ('survived').
*   **Missing Values:** 'age' (177 missing), 'embarked' (2 missing), 'deck' (688 missing), 'embark_town' (2 missing), 'cabin' (687 missing).

### 3. Algorithms Used

**California Housing Dataset:**
*   **Algorithm:** Linear Regression (a supervised learning algorithm for predicting a continuous target variable).

**Titanic Dataset:**
*   **Algorithm:** Logistic Regression (a supervised learning algorithm suitable for binary classification problems).

### 4. Model Performance

**California Housing Dataset (Linear Regression):**
*   **Metric:** Root Mean Squared Error (RMSE).
*   **Result:** RMSE = 0.7456.
*   **Interpretation:** On average, the model's predictions for median house value (in hundreds of thousands of dollars) differed from the actual values by approximately $74,560.

**Titanic Dataset (Logistic Regression):**
*   **Metric:** Accuracy Score and Classification Report.
*   **Result:** Accuracy = 0.7821.
*   **Interpretation:** The model correctly predicted the survival outcome for approximately 78.21% of the passengers in the test set. The classification report showed good precision, recall, and F1-scores for both 'survived' (class 1) and 'not survived' (class 0).

### 5. Challenges Faced

**Titanic Dataset:**
*   **Missing Data:** Significant missing values in 'age', 'embarked', and especially 'deck' required preprocessing. 'age' and 'embarked' were imputed, while 'deck' was dropped due to the high percentage of missing data.
*   **Categorical Features:** Conversion of categorical features (`sex`, `embarked`) into numerical format using one-hot encoding was necessary for the Logistic Regression model.
*   **Feature Selection:** Deciding which columns to drop (e.g., 'name', 'ticket', 'cabin', 'alive', 'adult_male', 'who', 'alone', 'embark_town', 'pclass', 'sibsp', 'parch') to avoid redundancy or use less informative features was a key preprocessing step.

### 6. Key Findings

**California Housing Dataset:**
*   The Linear Regression model provides a reasonable baseline for predicting median house values, with an average error of about $74,560. Further analysis could explore feature engineering, polynomial features, or more complex regression models to improve performance.

**Titanic Dataset:**
*   Even with basic data cleaning and a simple Logistic Regression model, a relatively high accuracy of 78.21% was achieved for predicting survival. This suggests that features like age, sex, and embarkation point are significant predictors of survival. More advanced models or extensive feature engineering could potentially yield higher accuracy.

```markdown
# LinkedIn Post Draft

This is week 4 of my internship at AnalystLab Africa, and what an insightful journey it's been! I've been diving deep into data analysis and machine learning, tackling two exciting projects:

1.  **California Housing Price Prediction (Regression)**:
    *   **Dataset**: Utilized the California Housing dataset to predict median house values.
    *   **Learnings**: Gained hands-on experience with feature selection, data splitting, training a Linear Regression model, and evaluating its performance using RMSE.
    *   **Performance**: Achieved a Root Mean Squared Error (RMSE) of **0.7456**, meaning predictions were, on average, off by approximately $74,560.

2.  **Titanic Survival Prediction (Classification)**:
    *   **Dataset**: Worked with the classic Titanic dataset to predict passenger survival.
    *   **Learnings**: Mastered crucial data cleaning techniques like handling missing values (imputing 'age' and 'embarked', dropping 'deck' due to sparsity) and converting categorical features to numerical using one-hot encoding for a Logistic Regression model. Also, learned to interpret classification reports.
    *   **Performance**: The Logistic Regression model yielded an impressive Accuracy Score of **0.7821**, correctly predicting survival for about 78.21% of the test set.

These projects have significantly enhanced my understanding of real-world data challenges and practical model implementation. Excited to see what the coming weeks hold!

#DataScience #MachineLearning #Internship #AnalystLabAfrica #Regression #Classification #Python #ScikitLearn #Pandas
```